### ISSUE TITLE
useEffect not run after change value in first useEffect

### ISSUE THREAD

**Issue by violin-wang** [testing-library/react-hooks-testing-library#102]
- `react-hooks-testing-library` version: 0.5.1
- `react-test-renderer` version: 16.8.6
- `react` version: 16.8.6
- `node` version: 11.1
- `npm` (or `yarn`) version: 1.15.2

### Relevant code or config:

```js
// useHello.js

import { useEffect, useState } from "react";

export default function useTest() {
  const [testUpdated, setTestUpdated] = useState(null);

  useEffect(() => {
    setTestUpdated({});
  }, [setTestUpdated]);

  useEffect(() => {
    console.log("testUpdated updated:", testUpdated);
  }, [testUpdated]);

  return {
    testUpdated
  };
}

```

```js
// useHello.test.js

import { renderHook } from "react-hooks-testing-library";

import useHello from "../useHello";

describe("useTest", () => {
  it("should be ok", async () => {
    const result = renderHook(useHello).result.current;

    expect(result).toEqual(expect.any(Object));
  });
});

```

### What you did:

I want to track `testUpdated` change. I change the `testUpdated` value in first `useEffect`.

### What happened:

It should output `testUpdated updated` twice, But now only once.

### Reproduction:

[Codesandbox Link](https://codesandbox.io/s/react-testing-library-demo-vu9r4?from-embed)

### Problem description:

It works well at 0.4.x version, but when I upgrade to 0.5.x, It's not work.

### Suggested solution:

I try to use `@testing-library/react` to test, And It's work fine.

**Comment by mpeyper** [comment]
That sanbox seems to be working fine to me?

![image](https://user-images.githubusercontent.com/23029903/59842326-b061f580-9399-11e9-870d-e478d0a807c0.png)

Am I missing something?

I don't think it's going to have any effect on this test, but your tests are going to be a bit more robust if you don't store the `result.current` value in a variable as it get's updated each time the component renders (like when your `setTestUpdated` get's called in `useEffect`.  The better way to write this test would be:

```js
describe("useTest", () => {
  it("should be ok", async () => {
    const { result } = renderHook(useHello);

    expect(result.current).toEqual(expect.any(Object));
  });
});
```

**Comment by violin-wang** [comment]
The test always be ok, but should output `testUpdated updated:` twice I think. I change to not store  `result.current` in local. But still no effect. So I change the [sanbox](https://codesandbox.io/embed/react-testing-library-demo-vu9r4), then the test will fail. `useHook` should return `{"foo":"bar"}` , but only return null. But in browser tab, it can display "{"foo":"bar"}".

**Comment by BabkinAleksandr** [comment]
@violin-wang I think you should just use `rerender`, because `afterChanged` will be updated on the next render

![image](https://user-images.githubusercontent.com/25172855/60499055-685ba080-9cc0-11e9-82ea-0e76ead02523.png)

**Comment by violin-wang** [comment]
@BabkinAleksandr  Thanks.. It works fine. But we can keep the same effect without `rerender` before the version upgrade (use 0.4.x). I am not sure if it's a right behavior.

**Comment by mpeyper** [comment]
v5.x.x was when we replaced `react-testing-library` (which uses `react-dom` under the hood) with `react-test-renderer`.  This made me wonder if the different renderers had different behaviour with respect to the `act` function.  Turns our it does, with the former seemingly doing a much better job at flushing effects before returning.  I took your codesandbox and [made one](https://codesandbox.io/s/useeffect-with-react-test-renderer-repro-64guc?fontsize=14&module=%2Fsrc%2Fexample.test.js&previewwindow=tests) that hopefully shows the difference in behaviour between the different renderers (I also tried just `react-dom` in case there was some magic RTL was using ever the top of it).  

I believe this behaviour is incorrect and I have [raised a bug in the react repo](https://github.com/facebook/react/issues/16034).  They may very well turn round and say it is working as intended, in which case there is not much I can do to help in this case (there are other benefits we gained from moving to `react-test-renderer` so we are unlikely to go back to `react-dom` or RTL).

**Comment by mpeyper** [comment]
Hi @violin-wang,

The react team got back to me on [the issue I raised](https://github.com/facebook/react/issues/16034#issuecomment-507903657) over there.

They've fixed it in the `16.9.0-alpha.0` version, so if you are able to use alphas, simply updating the version should see your tests go green.

If you can't use alphas, I recommend using the `rerender` work around from @BabkinAleksandr as it shouldn't break if the new version is officially released (although it will be unecessary) so you can move on with relative piece of mind.

**Comment by violin-wang** [comment]
@mpeyper  Thank you for telling me the problems.. I will use `v0.4.x` version until react stable version fix it.
